### ISSUE TITLE
Too many links

### ISSUE THREAD

**Issue by ionox0** [DataBiosphere/toil#2682]
Hello Toil team, was wondering if you would be able to comment on an issue we're experiencing.

```
s19.cbio.private 2019-05-20 14:17:29,783 MainThread WARNING toil.leader: Y/b/jobO9tOn0    Got workflow error
s19.cbio.private 2019-05-20 14:17:29,787 MainThread WARNING toil.leader: Y/b/jobO9tOn0    OSError: [Errno 31] Too many links
s19.cbio.private 2019-05-20 14:17:29,787 MainThread WARNING toil.leader: Y/b/jobO9tOn0    ERROR:cwltool:Got workflow error
s19.cbio.private 2019-05-20 14:17:29,791 MainThread WARNING toil.leader: Y/b/jobO9tOn0    OSError: [Errno 31] Too many links
s19.cbio.private 2019-05-20 14:17:29,796 MainThread WARNING toil.leader: Y/b/jobO9tOn0    ERROR:toil.worker:Exiting the worker because of a failed job on host
```

I assume we're dealing with this issue:
https://serverfault.com/questions/360961/what-does-the-message-too-many-links-mean-when-using-cp-on-linux

Is there any information regarding the type of filesystem that was used for the 20,000 RNAseq samples run using Toil? (ext3 vs ext4)

We are trying to run ~400 pairs of samples for this run, so perhaps this is above the limit for an ext3 filesystem using Toil. Thank you for any information you could provide.

┆Issue is synchronized with this [Jira Task](https://ucsc-cgl.atlassian.net/browse/TOIL-365)
┆Issue Number: TOIL-365

**Comment by adamnovak** [comment]
What version of Toil are you using, exactly? And are you running with caching enabled? And what job store are you using (presumably the filesystem-based one)?

The ext4 limit on links is ~65k, so you wouldn't expect to hit it with 400 samples.

It could be that there are too many hard links to files in the cache (either too many jobs are trying to use them, or links from old jobs are not being cleaned up properly), or it could be that we're trying to over-fill a directory somewhere. I've recently been poking around in the caching code, and depending on what version you have I'd want to look at my recent changes first.

I think the 20,000 sample RNAseq workflow was run using the AWS-based jobstore, and not a filesystem-based one at all. As for what filesystem the nodes used, I think ext4. @jvivian would know for sure.

**Comment by diekhans** [comment]
From my weak knowledge, one doesn't want to put a large number
of files in a ext3 directory, even if you don't hit the limit,
as it will be very slow.

ext4 is much better at speed for large directories, and XFS is
even faster and I don't *think* has a hard limit.

**Comment by jvivian** [comment]
@adamnovak — I'm almost certain it was ext4. This is corroborated by the code base for CGCloud, which we used before Toil got auto-scaling functionality merged in.

**Comment by ionox0** [comment]
Thank you @adamnovak for the information, we are using `--disableCaching`. We are also using the `toil-cwl-runner` command, and from the logs I can see that it is indeed the `NonCachingFileStore` that is being used. 

The version of Toil is https://pypi.org/project/toil-ionox0/0.0.1/ which I have forked from `Toil 3.15` and changed the version of the `subprocess32` requirement to avoid a dependency issue when installing Toil using `pip` (which may no longer be an issue). We would like to upgrade to 3.19 but have to deal with some incompatibilities due to changes in the length of names of tmp dirs. 

After checking the sizes of each directory in the filestore, there is nothing larger than the 32,000 limit, apart from the log directory (because we are using `--logDebug`):
```
./log 243958
```

Here is the full stack trace:
```
ERROR:cwltool:Got workflow error
Traceback (most recent call last):
  File "/home/johnsoni/virtualenvs/pipeline_StructuralVariants/lib/python2.7/site-packages/cwltool-1.0.20180306140409-py2.7.egg/cwltool/executors.py", line 98, in run_jobs
    for r in jobiter:
  File "/home/johnsoni/virtualenvs/pipeline_StructuralVariants/lib/python2.7/site-packages/cwltool-1.0.20180306140409-py2.7.egg/cwltool/command_line_tool.py", line 374, in job
    builder.pathmapper = self.makePathMapper(reffiles, builder.stagedir, **make_path_mapper_kwargs)
  File "/home/johnsoni/virtualenvs/pipeline_StructuralVariants/lib/python2.7/site-packages/toil_ionox0-0.0.1-py2.7.egg/toil/cwl/cwltoil.py", line 233, in makePathMapper
    get_file=kwargs["toil_get_file"])
  File "/home/johnsoni/virtualenvs/pipeline_StructuralVariants/lib/python2.7/site-packages/toil_ionox0-0.0.1-py2.7.egg/toil/cwl/cwltoil.py", line 196, in __init__
    stagedir, separateDirs=separateDirs)
  File "/home/johnsoni/virtualenvs/pipeline_StructuralVariants/lib/python2.7/site-packages/cwltool-1.0.20180306140409-py2.7.egg/cwltool/pathmapper.py", line 229, in __init__
    self.setup(dedup(referenced_files), basedir)
  File "/home/johnsoni/virtualenvs/pipeline_StructuralVariants/lib/python2.7/site-packages/cwltool-1.0.20180306140409-py2.7.egg/cwltool/pathmapper.py", line 282, in setup
    self.visit(fob, stagedir, basedir, copy=fob.get("writable"), staged=True)
  File "/home/johnsoni/virtualenvs/pipeline_StructuralVariants/lib/python2.7/site-packages/toil_ionox0-0.0.1-py2.7.egg/toil/cwl/cwltoil.py", line 211, in visit
    self.visitlisting(obj.get("listing", []), tgt, basedir, copy=copy, staged=staged)
  File "/home/johnsoni/virtualenvs/pipeline_StructuralVariants/lib/python2.7/site-packages/cwltool-1.0.20180306140409-py2.7.egg/cwltool/pathmapper.py", line 234, in visitlisting
    self.visit(ld, stagedir, basedir, copy=ld.get("writable", copy), staged=staged)
  File "/home/johnsoni/virtualenvs/pipeline_StructuralVariants/lib/python2.7/site-packages/toil_ionox0-0.0.1-py2.7.egg/toil/cwl/cwltoil.py", line 211, in visit
    self.visitlisting(obj.get("listing", []), tgt, basedir, copy=copy, staged=staged)
  File "/home/johnsoni/virtualenvs/pipeline_StructuralVariants/lib/python2.7/site-packages/cwltool-1.0.20180306140409-py2.7.egg/cwltool/pathmapper.py", line 234, in visitlisting
    self.visit(ld, stagedir, basedir, copy=ld.get("writable", copy), staged=staged)
  File "/home/johnsoni/virtualenvs/pipeline_StructuralVariants/lib/python2.7/site-packages/toil_ionox0-0.0.1-py2.7.egg/toil/cwl/cwltoil.py", line 211, in visit
    self.visitlisting(obj.get("listing", []), tgt, basedir, copy=copy, staged=staged)
  File "/home/johnsoni/virtualenvs/pipeline_StructuralVariants/lib/python2.7/site-packages/cwltool-1.0.20180306140409-py2.7.egg/cwltool/pathmapper.py", line 234, in visitlisting
    self.visit(ld, stagedir, basedir, copy=ld.get("writable", copy), staged=staged)
  File "/home/johnsoni/virtualenvs/pipeline_StructuralVariants/lib/python2.7/site-packages/toil_ionox0-0.0.1-py2.7.egg/toil/cwl/cwltoil.py", line 217, in visit
    resolved = self.get_file(loc) if self.get_file else loc
  File "/home/johnsoni/virtualenvs/pipeline_StructuralVariants/lib/python2.7/site-packages/toil_ionox0-0.0.1-py2.7.egg/toil/cwl/cwltoil.py", line 268, in toilGetFile
    srcPath = fileStore.readGlobalFile(fileStoreID[7:])
  File "/home/johnsoni/virtualenvs/pipeline_StructuralVariants/lib/python2.7/site-packages/toil_ionox0-0.0.1-py2.7.egg/toil/fileStore.py", line 1659, in readGlobalFile
    self.jobStore.readFile(fileStoreID, localFilePath, symlink=symlink)
  File "/home/johnsoni/virtualenvs/pipeline_StructuralVariants/lib/python2.7/site-packages/toil_ionox0-0.0.1-py2.7.egg/toil/jobStores/fileJobStore.py", line 309, in readFile
    os.link(jobStoreFilePath, localFilePath)
OSError: [Errno 31] Too many links
```

I will ask our HPC team what they would expect the limits to be on our filesystems and respond. Thank you very much for the information.

**Comment by adamnovak** [comment]
To me this looks like maybe cwltool is trying to read loads and loads
of files into one directory, or one file into loads and loads of
directories.

I think a possible solution for this issue is probably to catch this
error at that point in the FileJobStore and try a copy instead of a
hardlink. That will handle the case of a very popular file, although
it won't do any good in the case of an over-full destination
directory. If the underlying cause is really the workflow over-filling
a directory, that might be a problem with the workflow, or with how
cwltool is trying to run it.

**Comment by ionox0** [comment]
That's possible, we are pairing each of the 400 samples with a single other sample, and scattering over both lists of samples. This may mean that one file is getting too many links created. I suppose we should be scattering only over the first input, and using a single File object for its pair.

A simple test does show that trying to create too many hardlinks to the same file errors on our filesystem:

```
$ touch asdf
$ for i in {1..1600}; do ln asdf $i; done
ln: creating hard link to `asdf': Too many links
ln: creating hard link to `asdf': Too many links
ln: creating hard link to `asdf': Too many links
ln: creating hard link to `asdf': Too many links
ln: creating hard link to `asdf': Too many links
ln: creating hard link to `asdf': Too many links
ln: creating hard link to `asdf': Too many links
ln: creating hard link to `asdf': Too many links
ln: creating hard link to `asdf': Too many links
...
```

**Comment by ionox0** [comment]
The reason we are using a single sample like so is so that we can support flexible pairing of samples, while still scattering over two lists. For example if some samples will be paired with the same sample, but others will be unique. In this case we would probably make the second list a single `File` and not scatter over it, but we'd like it to be able to support both without having to change the cwl code. Thank you again for all your help!

**Comment by ionox0** [comment]
There's no limit for softlinks though, so perhaps that could be an alternative within cwltool

**Comment by adamnovak** [comment]
OK, that looks like we ought to be able to solve it by just falling
back to a copy when we can't make a hard link.

**Comment by adamnovak** [comment]
I've been wanting to revise the whole copying/linking/symlinking
decision system, but that is blocked on redoing the caching system to
be less opinionated about the link count of files. So I don't know if
I can lay all the groundwork for haw workflows ought to communicate
that symlinks are OK all the way back to the job store to let symlinks
be used here.

**Comment by DailyDreaming** [comment]
Stale, closing.  I'm guessing this was fixed by Adam's changes since he has since revised the whole copying/linking/symlinking decision system, but please reopen if this is still an issue.
